File: argus_synchro/ClsMMap.py
"""---------------------------
共有メモリ管理クラス
---------------------------"""

from __future__ import annotations

import mmap
import os
import struct

import numpy as np
from numpy.typing import NDArray

from argus_synchro.common.app_logger import AppLogger, AppLoggerFactory


class ByteSize:
    UNIT = 1
    INT8 = 1  # Char
    INT16 = 2  # Short
    INT32 = 4  # long
    INT64 = 8  # longlong
    FLOAT = 4
    MAP_ALL = 10 * 1024 * 1024
    # MAP_ALL = 10 * 1024 * 15  # damp時に巨大すぎないような一時指定


class classMMap:
    def __init__(self, mapFilePath: str, app_logger_factory: AppLoggerFactory) -> None:
        self._logger: AppLogger = app_logger_factory.register_from_type(self.__class__)
        # self.endian = sys.byteorder
        self.endian = "little"
        self.UNIT_SIZE = ByteSize.UNIT
        self.mapFilePath: str = mapFilePath
        self.size: int = ByteSize.MAP_ALL
        self._fp: object | None = None
        self._mm: mmap.mmap | None = None

        # ファイルをサイズ固定で作成／補正
        self._ensure_file(self.mapFilePath, self.size)
        # マップを開く（書込可）
        self._open_map()

        self._logger.info("__init__ comp.")

    # ---------- internal ----------
    def _ensure_file(self, path: str, size: int) -> None:
        # 既存サイズと違う場合も truncate で合わせる
        fd = os.open(path, os.O_RDWR | os.O_CREAT)
        try:
            cur = os.path.getsize(path)
            if cur != size:
                os.ftruncate(fd, size)
        finally:
            os.close(fd)
        self._logger.info(f"_ensure_file ok: {path} ({size} bytes)")

    def _open_map(self) -> None:
        # Windows での共有を考慮した変更.
        self._fp = open(self.mapFilePath, "r+b", buffering=0)
        # 明示サイズ＋書込可でマップ（length=0は避ける）
        self._mm = mmap.mmap(
            self._fp.fileno(), length=self.size, access=mmap.ACCESS_WRITE
        )
        self._mm.seek(0)
        self._logger.info("_open_map fin.")

    # ---------- public ----------
    def flush(self) -> None:
        if self._mm:
            self._mm.flush()

    def dispose(self) -> None:
        try:
            if self._mm:
                self._mm.flush()
                self._mm.close()
        finally:
            self._mm = None
            if self._fp:
                self._fp.close()
                self._fp = None
        self._logger.info("dispose fin.")

    # ---------- read / write ----------
    def ReadInt(self, adr: int, byte_size: int, sign: bool) -> int | None:
        try:
            assert self._mm is not None
            self._mm.seek(adr)
            bytes: bytes = self._mm.read(byte_size)
            val: int = int.from_bytes(bytes, self.endian, signed=sign)
            self._mm.seek(0)
            return val
        except Exception as e:
            self._logger.info("ReadInt except: " + str(e).replace("\n", ""))
            return None

    def ReadInt64(self, adr: int) -> int | None:
        return self.ReadInt(adr, ByteSize.INT64, False)

    def ReadInt32(self, adr: int) -> int | None:
        return self.ReadInt(adr, ByteSize.INT32, False)

    def ReadInt16(self, adr: int) -> int | None:
        return self.ReadInt(adr, ByteSize.INT16, False)

    def ReadInt8(self, adr: int) -> int | None:
        return self.ReadInt(adr, ByteSize.INT8, False)

    def WriteInt(self, adr: int, data: int, byte_size: int, sign: bool) -> None:
        try:
            assert self._mm is not None
            # 正しい範囲チェック（符号あり／なし）
            if sign:
                min_v = -(1 << (8 * byte_size - 1))
                max_v = (1 << (8 * byte_size - 1)) - 1
            else:
                min_v = 0
                max_v = (1 << (8 * byte_size)) - 1
            if not (min_v <= int(data) <= max_v):
                self._logger.info(
                    f"Error. Out of range for {byte_size}B sign={sign}: {data}"
                )
                return

            bytes: bytes = int(data).to_bytes(byte_size, self.endian, signed=sign)
            self._mm[adr : adr + byte_size] = bytes
        except Exception as e:
            self._logger.info("WriteInt except " + str(e).replace("\n", ""))

    def WriteInt64(self, adr: int, data: int) -> None:
        self.WriteInt(adr, data, ByteSize.INT64, False)

    def WriteInt32(self, adr: int, data: int) -> None:
        self.WriteInt(adr, data, ByteSize.INT32, False)

    def WriteInt16(self, adr: int, data: int) -> None:
        self.WriteInt(adr, data, ByteSize.INT16, False)

    def WriteSignedInt16(self, adr: int, data: int) -> None:
        self.WriteInt(adr, data, ByteSize.INT16, True)

    def WriteInt8(self, adr: int, data: int) -> None:
        self.WriteInt(adr, data, ByteSize.INT8, False)

    def WriteFloat(self, adr: int, data: float) -> None:
        try:
            assert self._mm is not None
            struct.pack_into("<f", self._mm, adr, float(data))
        except Exception as e:
            self._logger.info("WriteFloat except " + str(e).replace("\n", ""))

    def ReadFloat(self, adr: int) -> float | None:
        try:
            assert self._mm is not None
            return struct.unpack_from("<f", self._mm, adr)[0]
        except Exception as e:
            self._logger.info("ReadFloat except " + str(e).replace("\n", ""))
            return None

    # 配列データ書き込み & Byteデータ直接書き込み.(JPEG画像書き込み等に使用)
    def WriteBytes(
        self, adr: int, data: NDArray[np.uint8] | bytes | bytearray | memoryview
    ) -> None:
        try:
            assert self._mm is not None
            mv: memoryview[int] = memoryview(data)
            self._mm[adr : adr + len(mv)] = mv
        except Exception as e:
            self._logger.info("WriteBytes except " + str(e).replace("\n", ""))

    def ReadBytes(self, adr: int, length: int) -> bytes | None:
        try:
            assert self._mm is not None
            return self._mm[adr : adr + length]
        except Exception as e:
            self._logger.info("ReadBytes except " + str(e).replace("\n", ""))
            return None

File: argus_synchro_lib/cpp/src/clsmmap/ClsMMap.cpp
#include "clsmmap/ClsMMap.h"

#include <fcntl.h>
#include <filesystem>
#include <fstream>
#include <optional>
#include <sys/mman.h>
#include <unistd.h>

#include <iostream>
#include <chrono>

// 実行環境を取得
static inline std::string getPlatform(void)
{
#if defined(_WIN32)
    std::string system = "Windows";
#elif defined(__linux__)
    std::string system = "Linux";
#elif defined(__APPLE__)
    std::string system = "Darwin";
#else
    std::string system = "Unknown OS";
#endif
    return system;
}

ByteOrder GetSystemByteOrder()
{
    int n = 1;
    // little endian if true
    if (*(char*)&n == 1)
    {
        return ByteOrder::LITTLE;
    }
    return ByteOrder::BIG;
}

classMMap::classMMap(const MMapConfig& cfg, const LoggerFunc logfunc)
    : cfg_(cfg), system_endian_(GetSystemByteOrder()), current_index_(0), logger_(PyLogger(logfunc))
{
    this->logger_.info("platform: %s", getPlatform().c_str());
    this->logger_.info("endian: %s", (bool(this->system_endian_) ? "little" : "big"));

    if (cfg_.paths.empty())
    {
        throw std::runtime_error("classMMap: dat_paths is empty.");
    }
    if (cfg_.map_size_bytes == 0)
    {
        throw std::runtime_error("classMMap: map_size_bytes must be > 0.");
    }
    ensure_and_map_all();
}

classMMap::~classMMap()
{
    this->dispose();
}

void classMMap::begin_frame() const
{
    this->WriteInt8(cfg_.proto.IsWriting_ADR, 1);
}

void classMMap::end_frame(bool is_initial)
{
    bool isRead = false;
    if (auto v = ReadInt8(cfg_.proto.IsReading_ADR))
    {
        isRead = (v.value() > 0);
    }
    if (isRead || is_initial)
    {
        WriteInt8(cfg_.proto.IsWriting_ADR, 0);
        current_index_ = (current_index_ + 1) % mms_.size();
        this->logger_.info("MMAP Index changed!");
    }
}

void classMMap::dispose()
{
    unmap_all();
}

std::optional<uint64_t> classMMap::ReadInt64(int adr) const
{
    std::optional<uint64_t> val = this->ReadInt<uint64_t>(adr, false);
    if (!val.has_value())
    {
        return val;
    }
    if (cfg_.endian != this->system_endian_)
    {
        // system内のendianと使用したいendianが異なっていたらswap
        val = __builtin_bswap64(val.value());
    }
    return val;
}

std::optional<uint32_t> classMMap::ReadInt32(int adr) const
{
    std::optional<uint32_t> val = this->ReadInt<uint32_t>(adr, false);
    if (!val.has_value())
    {
        return val;
    }
    if (cfg_.endian != this->system_endian_)
    {
        // system内のendianと使用したいendianが異なっていたらswap
        val = __builtin_bswap32(val.value());
    }
    return val;
};

std::optional<uint16_t> classMMap::ReadInt16(int adr) const
{
    std::optional<uint16_t> val = this->ReadInt<uint16_t>(adr, false);
    if (!val.has_value())
    {
        return val;
    }
    if (cfg_.endian != this->system_endian_)
    {
        // system内のendianと使用したいendianが異なっていたらswap
        val = __builtin_bswap16(val.value());
    }
    return val;
};

std::optional<int16_t> classMMap::ReadSignedInt16(int adr) const
{
    std::optional<int16_t> val = this->ReadInt<int16_t>(adr, false);
    if (!val.has_value())
    {
        return val;
    }
    if (cfg_.endian != this->system_endian_)
    {
        // system内のendianと使用したいendianが異なっていたらswap
        val = __builtin_bswap16(val.value());
    }
    return val;
}

std::optional<uint8_t> classMMap::ReadInt8(int adr) const
{
    return this->ReadInt<uint8_t>(adr, false);
};

std::optional<float> classMMap::ReadFloat(int adr) const
{
    try
    {
        this->isValidMemoryMappingAndAddress(adr, sizeof(float));
        float val = 0.0f;
        std::memcpy(&val, current_mm() + adr, sizeof(float));
        return val;
    }
    catch (const std::exception& e)
    {
        // try catch
        std::cerr << "ReadFloat except " << e.what() << std::endl;
        return std::nullopt;
    }
}

void classMMap::WriteInt64(int adr, uint64_t data) const
{
    if (cfg_.endian != this->system_endian_)
    {
        // system内のendianと使用したいendianが異なっていたらswap
        data = __builtin_bswap64(data);
    }
    this->WriteInt<uint64_t>(adr, data, false);
};

void classMMap::WriteInt32(int adr, uint32_t data) const
{
    if (cfg_.endian != this->system_endian_)
    {
        // system内のendianと使用したいendianが異なっていたらswap
        data = __builtin_bswap32(data);
    }
    this->WriteInt<uint32_t>(adr, data, false);
};

void classMMap::WriteInt16(int adr, uint16_t data) const
{
    if (cfg_.endian != this->system_endian_)
    {
        // system内のendianと使用したいendianが異なっていたらswap
        data = __builtin_bswap16(data);
    }
    this->WriteInt<uint16_t>(adr, data, false);
};

void classMMap::WriteSignedInt16(int adr, int16_t data) const
{
    if (cfg_.endian != this->system_endian_)
    {
        // system内のendianと使用したいendianが異なっていたらswap
        data = __builtin_bswap16(data);
    }
    this->WriteInt<int16_t>(adr, data, true);
};

void classMMap::WriteInt8(int adr, uint8_t data) const
{
    this->WriteInt<uint8_t>(adr, data, false);
};

void classMMap::WriteFloat(int adr, float data) const
{
    try
    {
        this->isValidMemoryMappingAndAddress(adr, sizeof(float));

        std::memcpy(current_mm() + adr, &data, sizeof(float));
    }
    catch (const std::exception& e)
    {
        // try catch
        std::cerr << "WriteFloat except " << e.what() << std::endl;
    }
}

void classMMap::WriteBytes(int adr, const Eigen::Ref<const Eigen::Matrix<uint8_t, Eigen::Dynamic, 1>>& data) const
{
    try
    {
        this->isValidMemoryMappingAndAddress(adr, data.size());
        std::memcpy(current_mm() + adr, data.data(), data.size());
    }
    catch (const std::exception& e)
    {
        // try catch
        std::cerr << "WriteBytes except " << e.what() << std::endl;
    }
};

void classMMap::WriteBytes(int adr, const std::vector<unsigned char>& data) const
{
    try
    {
        this->isValidMemoryMappingAndAddress(adr, data.size());
        std::memcpy(current_mm() + adr, data.data(), data.size());
    }
    catch (const std::exception& e)
    {
        // try catch
        std::cerr << "WriteBytes except " << e.what() << std::endl;
    }
};

size_t classMMap::get_file_index() const
{
    return cfg_.paths.size();
}

std::optional<Eigen::Matrix<uint8_t, Eigen::Dynamic, 1>> classMMap::ReadBytes(size_t index, int adr, int length) const
{
    try
    {
        if (index >= mms_.size() || mms_[index] == nullptr || fds_[index] == -1)
        {
            throw std::out_of_range("ReadBytes: invalid buffer index");
        }
        if (adr < 0 || static_cast<size_t>(adr) >= cfg_.map_size_bytes)
        {
            throw std::out_of_range("ReadBytes: adr out of range");
        }

        const char* base = mms_[index];

        if (length < 0)
        {
            const int start = cfg_.proto.Start_ADR;
            const size_t total = cfg_.map_size_bytes;
            Eigen::Matrix<uint8_t, Eigen::Dynamic, 1> val(total);

            const size_t tail = total - static_cast<size_t>(start);
            std::memcpy(val.data(), base + start, tail);
            if (start > 0)
            {
                std::memcpy(val.data() + tail, base, static_cast<size_t>(start));
            }
            return val;
        }

        // 部分読み出し
        if (length < 0 || static_cast<size_t>(adr + length) > cfg_.map_size_bytes)
        {
            throw std::out_of_range("ReadBytes: length out of range");
        }
        Eigen::Matrix<uint8_t, Eigen::Dynamic, 1> val(length);
        std::memcpy(val.data(), base + adr, static_cast<size_t>(length));
        return val;
    }
    catch (const std::exception& e)
    {
        std::cerr << "ReadBytes(indexed) except " << e.what() << std::endl;
        return std::nullopt;
    }
}

// fileに明示的に書き込むための関数
void classMMap::flush() const
{
    if (current_mm() && msync(current_mm(), cfg_.map_size_bytes, MS_SYNC) == -1)
    {
        std::cerr << "msync failed" << std::endl;
    }
}

void classMMap::ensure_and_map_all()
{
    const auto& paths = cfg_.paths;
    fds_.resize(paths.size(), -1);
    mms_.resize(paths.size(), nullptr);

    for (size_t i = 0; i < paths.size(); ++i)
    {
        const std::string& path = paths[i];

        if (std::filesystem::exists(path))
        { // 前回異常終了対策でファイル削除
            std::filesystem::remove(path);
        }

        int fd = ::open(path.c_str(), O_RDWR | O_CREAT, 0666);
        if (fd == -1)
        {
            throw std::runtime_error("Failed to open or create file: " + path);
        }

        try
        {
            // サイズが異なるなら ftruncate
            this->logger_.info("dataPathIndex: {%d}, {%s]}", i, path.c_str());
            if (std::filesystem::exists(path))
            {
                const auto cur = static_cast<int64_t>(std::filesystem::file_size(path));
                if (static_cast<size_t>(cur) != cfg_.map_size_bytes)
                {
                    if (ftruncate(fd, static_cast<off_t>(cfg_.map_size_bytes)) == -1)
                    {
                        ::close(fd);
                        throw std::runtime_error("Failed to set file size: " + path);
                    }
                }
            }
            else
            {
                if (ftruncate(fd, static_cast<off_t>(cfg_.map_size_bytes)) == -1)
                {
                    ::close(fd);
                    throw std::runtime_error("Failed to set file size: " + path);
                }
            }
            this->logger_.info("classMMap %s created", path.c_str());

            void* mm = ::mmap(nullptr, cfg_.map_size_bytes, PROT_READ | PROT_WRITE, MAP_SHARED, fd, 0);
            if (mm == MAP_FAILED)
            {
                ::close(fd);
                throw std::runtime_error("Failed to mmap the file: " + path);
            }

            fds_[i] = fd;
            mms_[i] = static_cast<char*>(mm);
        }
        catch (...)
        {
            ::close(fd);
            throw;
        }
    }
}

void classMMap::unmap_all()
{
    for (size_t i = 0; i < mms_.size(); ++i)
    {
        if (mms_[i])
        {
            ::munmap(mms_[i], cfg_.map_size_bytes);
            mms_[i] = nullptr;
        }
        if (fds_[i] != -1)
        {
            ::close(fds_[i]);
            fds_[i] = -1;
        }
    }
    fds_.clear();
    mms_.clear();
}

void classMMap::isValidMemoryMappingAndAddress(int adr, size_t size) const
{
    if (mms_.empty() || current_fd() == -1 || !current_mm())
    {
        throw std::runtime_error("Memory not initialized");
    }
    if (adr < 0 || size > (cfg_.map_size_bytes - static_cast<size_t>(adr)))
    {
        throw std::out_of_range("Out-of-range address");
    }
}

template <typename T> std::optional<T> classMMap::ReadInt(int adr, bool /*is_signed*/) const
{
    try
    {
        isValidMemoryMappingAndAddress(adr, sizeof(T));
        T v{};
        std::memcpy(&v, current_mm() + adr, sizeof(T));
        return v;
    }
    catch (const std::exception& e)
    {
        std::cerr << "ReadIntT except: " << e.what() << std::endl;
        return std::nullopt;
    }
}

template <typename T> void classMMap::WriteInt(int adr, T data, bool /*is_signed*/) const
{
    try
    {
        isValidMemoryMappingAndAddress(adr, sizeof(T));
        std::memcpy(current_mm() + adr, &data, sizeof(T));
    }
    catch (const std::exception& e)
    {
        std::cerr << "WriteIntT except: " << e.what() << std::endl;
    }
}